### DISCUSSION TITLE
Unable to Download the video

### DISCUSSION THREAD

**Discussion by thetechank** [aclap-dev/video-downloadhelper#208] [Help and Q&A]
### Your question

Hello,
When i click on the download for the available videos I get the errro as seen in the screenshot below. Please help
![Screenshot 2024-02-23 at 15 37 36](https://github.com/aclap-dev/video-downloadhelper/assets/36125875/c9fc3116-322c-4d27-8ea8-b53b0f91e6ec)

### Addon & CoApp version and details

version: 8.2.1.1
channel: stable
build date: 2024-02-19
build options: linuxlic: false, noyt: true, browser: chrome
lang: en-GB
license: accepted
key (only 16 first characters): 0b3dd1db8b0348aa
platform: arm64 mac
UA: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36
coapp: yes. Info:
{
  "id": "net.downloadhelper.coapp",
  "name": "VdhCoApp",
  "version": "2.0.19",
  "binary": "/Applications/net.downloadhelper.coapp.app/Contents/MacOS/vdhcoapp",
  "displayName": "VdhCoApp",
  "description": "Video DownloadHelper companion app",
  "target": {
    "os": "mac",
    "arch": "arm64",
    "node": "18"
  },
  "home": "/Users/ank",
  "converterBinary": "/Applications/net.downloadhelper.coapp.app/Contents/MacOS/ffmpeg",
  "converterBase": "ffmpeg",
  "converterBaseVersion": "n6.1-11-gc30f360955"
}

**Comment by jcfrog** [comment]
Hi,
Could you share a public link of a video with this issue so that the devs can reproduce it?

**Reply by thetechank** [reply]
[Here you go:](https://www.bridgewhy.com/s/preview/courses/BGP-from-Scratch-English-Version-1-650d91d0e4b08090e97a7bf2#650d60e4e4b0ec607e8ad5a1)

**Reply by GrampaWildWilly** [reply]
Thanks for that.  For the record, I am running Windows 11 Pro 64-bit version 23H2 build 22631.3155, Firefox 123.0 64-bit, licensed VDH 8.2.1.3a1 beta, CoApp 2.0.19.

My VDH session went like this:

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/8b9dbba1-b292-4df6-9084-38de1827fa2f)

That ended with this error:

```
Exit code: 1

moz-extension://9d469d39-7727-4f85-8beb-0761511221b9/background/main.js:1
receive@moz-extension://9d469d39-7727-4f85-8beb-0761511221b9/background/main.js:1:13157
Zp</callCatchAppNotFound/</<@moz-extension://9d469d39-7727-4f85-8beb-0761511221b9/background/main.js:1:42624
```

So . . .

![#02](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/56c8e2df-d693-47ef-92f8-e964c701ab60)

Each of those 3 attempts ended with the exact same error.

Time for Plan B.

![#03](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/add7614d-14c3-4553-b912-4b03fb056507)
![#04](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/2655b2b1-1812-43e0-8ab2-ecf011dc80e4)

I ran all 5 of those manifests through ffprobe:

[ffprobe Master.txt](https://github.com/aclap-dev/video-downloadhelper/files/14389407/ffprobe.Master.txt)
[ffprobe Audio.txt](https://github.com/aclap-dev/video-downloadhelper/files/14389410/ffprobe.Audio.txt)
[ffprobe 500k Video.txt](https://github.com/aclap-dev/video-downloadhelper/files/14389411/ffprobe.500k.Video.txt)
[ffprobe 1M Video.txt](https://github.com/aclap-dev/video-downloadhelper/files/14389412/ffprobe.1M.Video.txt)
[ffprobe 2M Video.txt](https://github.com/aclap-dev/video-downloadhelper/files/14389413/ffprobe.2M.Video.txt)

I then just scrolled through the Network Monitor.  I found exactly one json, which I saved to my system.  It did contain the URL of an m3u8.  These URLs make my eyes spin.  This URL might have been the same as one I had already gotten off one of the other manifests I had already run through ffprobe.  But I tried this one anyway:

[ffprobe json.txt](https://github.com/aclap-dev/video-downloadhelper/files/14389516/ffprobe.json.txt)

Not much difference.

If the site feeds errors to our software, it's rather unlikely that a download will succeed.

On the other hand, I did play the video in the web page & skipped around in it to sample it.  I'm not about to watch this all the way through for over an hour.  So it's possible that if there is corruption, I might have skipped over it.  Possible.  How likely?  I have no idea.

Maybe somebody who isn't pretending to be smart (like me) but who actually is smart (@paulrouget) has some deeper insights.  This one has me stumped.

**Comment by Mattmjs** [comment]
I think this example is using encryption. If you look at the response for the index and stream manifests you'll see a long garbled
string. It's the same even for the ts segments.

One of the stream manifests I downloaded has this inside it:

#EXT-X-KEY:METHOD=AES-128

I assume all the stream manifests have it. Am I right or wrong on this one. Also I have seen another case matching this one from the
google group forum if you want to check it out: 

https://groups.google.com/g/video-downloadhelper-q-and-a/c/0UzHIZ4pZEU

**Reply by GrampaWildWilly** [reply]
Well. there's the curious case of this one: #162.  The manifests in that one have AES-128 encryption, & after some fiddling with the stream manifest, I was able to download that one with ffmpeg.  But @paulrouget commented in there that there was something non-standard about the way the web site was presenting its content.  Apparently, that's what tripped up VDH in that case.  So the mere presence of AES-128 encryption does not, by itself, mean that something can't be downloaded.  The item in question very well may not be downloadable.  But the encryption is not necessarily a roadblock.  In the case of the video in this discussion that I tested, you can see that ffprobe said on all the manifests that there were chunks of the streams that contained invalid data.  The error messages don't specify exactly how the data is invalid.  It just said it is invalid.  I suspect this is another case of a web site using tools to encode their content in such a way that it doesn't fit within the array of structures that ffprobe understands.  It's invalid to ffprobe.  It's not invalid to the player on the web page, but it's invalid to ffprobe.  This is probably just more evidence of the fact that there is no ISO standard for web sites to adhere to when it comes time for them to present their A/V content.  I have said a number of times that I think it's something of a miracle that VDH works on as many web sites as it does.  It's probably just as miraculous that ffmpeg also seems to be widely able to get stuff.  From where I'm sitting, the site in this discussion seems to be one of those that these 2 tools can't handle.

You've shown your dexterity with yt-dlp.  How does it fare on this video?

**Reply by GrampaWildWilly** [reply]
Another comment on encryption.  If something is encrypted, that's not a problem if you have the decryption key.  Generally, the stream manifest will contain a reference to the key file like this:

#EXT-X-KEY:METHOD=AES-128,URI="URL of key file"

That URL may be complete, which is to say in the form https://whatever.  Or it may be just a partial URL which would require runtime processing to flesh it out to the full URL.  Whatever the case, that key file usually appears somewhere in the Network Monitor.  I have usually observed that the file name has an extension of .key.  You don't even have to look inside it.  You just need to copy its URL into the URI= parameter of the #EXT-X-KEY statement in the stream manifest.  I think when a site really is encrypting its content & doesn't want downloaders to work, the key file won't appear in the Network Monitor.  But if the key file does appear in the Network Monitor, then you should be able to download the item in question.

In the case of discussion #162, look inside the ffmpeg log file I posted there.  You can see where ffmpeg opens the key file 1080.key.  That happened to be the file name of the key file in that case.  It fishes some information out of the file.  I have no idea what format it's in.  I never bothered to actually inspect the contents of the key file.  But in every subsequent log line ffmpeg writes, it shows a URL with some part of it that says "key."  I suspect that each one of those shows us what it fished out of the key file.

This just means that the web site has to cooperate or we'll be blocked.  If we don't have the key file, we'll be stuck.  If we do have the key file, it will just work without any effort on our part.

I have to say that I see no evidence of encryption in the ffprobe logs I posted above.  There is no reference to "key" in any of those logs.  I don't think the issue with this video has anything to do with encryption.  In fact, I don't think this video is encrypted.

**Reply by GrampaWildWilly** [reply]
The strings you aptly call "long garbled" are not necessarily connected to encryption.  They are more likely just identifiers that associate your client browser session with resources on the server web site for the duration of your visit.  The server needs to know where you are in order to feed you the video in the player in the web page.  Those long garbled strings also typically contain expiration information.  The server isn't going to keep your session active indefinitely.  At some point, you're going to close that web page.  After some delay, short or long, the server is going to ditch the information it was keeping about you.  So, session & expiration information is most likely what those long garbled strings are, not decryption information.

**Reply by Mattmjs** [reply]
I'm no expert on these things , I'm just wondering what could be happening that downloading fails on this one.
I didn't see any key urls anywhere on this one compared to the other one. But I did try yt-dlp and was surprised to see a format
list come up.

[generic] Extracting URL: https://dnqe0v482iyb1.cloudfront.net/spees/w/o/6502ee39e4b035ccd352401d/v/650d60e4e4b0ec607e8ad5a...07e8ad5a1/index.m3u8
[generic] index: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] index: Downloading m3u8 information
[generic] index: Checking m3u8 live status
[info] Available formats for index:
ID          EXT RESOLUTION │   FILESIZE   TBR PROTO │ VCODEC        VBR ACODEC     MORE INFO
────────────────────────────────────────────────────────────────────────────────────────────
audio-0-eng mp4 audio only │                  m3u8  │ audio only        unknown    [eng] eng
650         mp4 852x480    │ ~358.56MiB  650k m3u8  │ avc1.64001f  650k video only
900         mp4 1280x720   │ ~496.47MiB  900k m3u8  │ avc1.64001f  900k video only
1500        mp4 1920x1080  │ ~827.45MiB 1500k m3u8  │ avc1.640028 1500k video only

But it quickly failed when attempting download :  ERROR: Incorrect AES key length (64 bytes)

**Reply by GrampaWildWilly** [reply]
Yeah, I have to retract what I said.  I looked at the manifests again & they do use AES-128 encryption.  The manifests say the file name of the key file is k/timestamp.  I found 2 files by that name in the Network Monitor.  Each one of them contains a string of unintelligible gobbledygook.  Perhaps not coincidentally, the gobbledygook happens to be 64 characters long.  The manifests do specify AES-128.  I am guessing (I don't really know anything about encryption) that AES-128 means the decryption key is supposed to be 128 bytes long.  The 2 objects named timestamp that I found in the Network Monitor had the exact same URL.  Kind of weird to have two objects at the same URL.  But their respective 64 bytes of content were different.  Maybe the trick is to create a single key file with the 2 strings of 64 bytes simply concatenated one after the other to make up 128 bytes.  Which one goes first?  Try one order & then the other, I suppose.  I already fiddled with a manifest over in #162.  It's your turn.  You fiddle with these manifests.  Plural.  The audio track seems to be separate from the video track here.  There is one manifest for the audio track & more than one manifest for the video track.  Seems a bit DASHy to me, despite the distinctly HLSy names with m3u8.  You'd have to fiddle the manifest for the audio track as well as one of the manifests for a video track.  Come on.  Take it as a challenge.

**Reply by Mattmjs** [reply]
I'm going to pass on that, it's far easier to screen record this one than to downloading. I see it as impossible.
But I'm not the one who wants this content, the original poster does.

**Reply by GrampaWildWilly** [reply]
Wise move.  I didn't pass on it.  But I failed miserably with this.  I couldn't get it to recognize my key file.  It kept inserting k/timestamp into the name of the key file, even though I had changed it to the name of a file on my system.  There is something somewhere in these manifests that has that k/timestamp name hard wired.  I couldn't get around it.  The approach I used in #162 just didn't work here.  It seems like the key file is in chunks just like the media streams are.  Seems weird to do that for such a small file but that's what it looks like.  Your screen recorder idea is probably the best way to go with this.

**Reply by thetechank** [reply]
I understand there is some issue with the video. The whole point of buying VDH was to avoid the screen recording. I posted this so that this could be solved but looks like it is not possible.

**Reply by jcfrog** [reply]
The dev team might have a look this week, thanks for your patience.
